src/auth/storage.js

```diff
@@ -1,53 +1,124 @@
-class MemoryStorage {
+import { USER_AUTH_KEY, REFRESH_TOKEN_KEY, DEVICE_ID_KEY, STATE_KEY } from './common';
+
+export class MemoryStorage {
   constructor() {
     this._data = {};
+    this._orderedKeys = [];
+    this.length = 0;
   }
 
   getItem(key) {
     return (key in this._data) ? this._data[key] : null;
   }
 
   setItem(key, value) {
+    this._orderedKeys.push(key);
     this._data[key] = value;
+    this.length++;
     return this._data[key];
   }
 
   removeItem(key) {
+    this._orderedKeys.pop(key);
     delete this._data[key];
+    this.length--;
```

Comment:
any reason we have to maintain `length` instead of just using `this._orderedKeys.length`?

Reply:
Yes– it is to fully implement the interface of the various other storage types: https://developer.mozilla.org/en-US/docs/Web/API/Storage/length . We need to give arbitrary accessors access to `.length` as a class member.

Reply:
OK, makes sense. Alternately you could add a class method that does this with `get length(){ return this._orderedKeys.length; }`